pbcommand/models/report.py

```diff
@@ -838,3 +838,252 @@ def _merge_tables(tables):
+                         [ColumnSpec.from_dict(c) for c in d['columns']])
+
+    def get_column_spec(self, id_):
+        return self._col_dict.get(id_, None)
+
+
+class PlotSpec(object):
+
+    def __init__(self, id_, description, caption, title, xlabel, ylabel):
+        self.id = id_
+        self.description = description
+        self.caption = caption
+        self.title = title
+        self.xlabel = xlabel
+        self.ylabel = ylabel
+
+    @staticmethod
+    def from_dict(d):
+        return PlotSpec(d['id'].split(".")[-1], d['description'],
+                        d['caption'], d['title'],
+                        d.get('xlabel', None), d.get('ylabel', None))
+
+
+class PlotGroupSpec(object):
+
+    def __init__(self, id_, title, description, legend, plots=()):
+        self.id = id_
+        self.title = title
+        self.description = description
+        self.legend = legend
+        self.plots = plots
+        self._plot_dict = {p.id: p for p in plots}
+
+    @staticmethod
+    def from_dict(d):
+        return PlotGroupSpec(d['id'].split(".")[-1], d['title'],
+                             d["description"], d['legend'],
+                             [PlotSpec.from_dict(p) for p in d['plots']])
+
+
+    def get_plot_spec(self, id_):
+        return self._plot_dict.get(id_, None)
+
+
+class ReportSpec(object):
+    """
+    Model for a specification of the expected content of a uniquely
+    identified report.  For obvious reasons this mirrors the Report model,
+    minus values and with added view metadata.  These specs should usually
+    be written out explicitly in JSON rather than built programatically.
+    """
+
+    def __init__(self, id_, version, title, description, attributes=(),
+                 plotgroups=(), tables=()):
+        self.id = id_
+        self.version = version
+        self.title = title
+        self.description = description
+        self.attributes = attributes
+        self.plotgroups = plotgroups
+        self.tables = tables
+        self._attr_dict = {a.id: a for a in attributes}
+        self._plotgrp_dict = {p.id: p for p in plotgroups}
+        self._table_dict = {t.id: t for t in tables}
+
+    @staticmethod
+    def from_dict(d):
+        return ReportSpec(d['id'], d['version'], d['title'], d['description'],
+                          [AttributeSpec.from_dict(a)
+                           for a in d['attributes']],
+                          [PlotGroupSpec.from_dict(p)
+                           for p in d['plotGroups']],
+                          [TableSpec.from_dict(t) for t in d['tables']])
+
+    def get_attribute_spec(self, id_):
+        return self._attr_dict.get(id_, None)
+
+    def get_plotgroup_spec(self, id_):
+        return self._plotgrp_dict.get(id_, None)
+
+    def get_table_spec(self, id_):
+        return self._table_dict.get(id_, None)
+
+    def validate_report(self, rpt, raise_if_found=True):
```

Comment:
The reason `validate_report` doesn't return the report is that I want to be able to test for errors in response to specific defects.